File: src/main/java/com/example/flightradar/OpenSkyClient.java
package com.example.flightradar;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Fetches live aircraft states from the OpenSky Network REST API ({@code /states/all}).
 *
 * <p>Used anonymously (no account / OAuth), which is the simplest path but is subject to stricter
 * rate limits. Any failure (HTTP 429/5xx, timeout, parse error) yields an empty list rather than an
 * exception, so the UI keeps running and simply skips that update.
 */
public final class OpenSkyClient {

    /** Bounding box (lamin, lomin, lamax, lomax) in degrees, used to scope and shrink the query. */
    public record BoundingBox(double laMin, double loMin, double laMax, double loMax) {}

    /** Roughly continental Europe — a sensible, traffic-dense default. */
    public static final BoundingBox EUROPE = new BoundingBox(35.0, -12.0, 60.0, 30.0);

    /** The whole planet. Heavier payload and higher rate-limit cost; use with care. */
    public static final BoundingBox WORLD = new BoundingBox(-90.0, -180.0, 90.0, 180.0);

    private static final String BASE_URL = "https://opensky-network.org/api/states/all";
    private static final String USER_AGENT = "flightradar-java/1.0 (personal demo)";

    private final HttpClient httpClient;
    private final ObjectMapper mapper = new ObjectMapper();
    private final BoundingBox boundingBox;

    public OpenSkyClient(BoundingBox boundingBox) {
        this.boundingBox = boundingBox;
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .build();
    }

    /**
     * Fetches the current aircraft states within the configured bounding box.
     *
     * @return the live states, or an empty list on any error (never {@code null})
     */
    public List<AircraftState> fetchStates() {
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(buildUrl()))
                    .header("User-Agent", USER_AGENT)
                    .timeout(Duration.ofSeconds(10))
                    .GET()
                    .build();

            HttpResponse<String> response =
                    httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                System.err.println("[OpenSky] HTTP " + response.statusCode()
                        + " — skipping this update.");
                return Collections.emptyList();
            }

            return parseStates(response.body());
        } catch (Exception e) {
            System.err.println("[OpenSky] request failed: " + e.getMessage()
                    + " — skipping this update.");
            return Collections.emptyList();
        }
    }

    private String buildUrl() {
        return BASE_URL
                + "?lamin=" + boundingBox.laMin()
                + "&lomin=" + boundingBox.loMin()
                + "&lamax=" + boundingBox.laMax()
                + "&lomax=" + boundingBox.loMax();
    }

    /**
     * Parses an OpenSky {@code /states/all} response body into aircraft states. Rows without a
     * usable position are skipped. Package-visible and static so it can be unit tested without HTTP.
     *
     * @param json the raw response body
     * @return the parsed states (never {@code null})
     */
    List<AircraftState> parseStates(String json) {
        List<AircraftState> result = new ArrayList<>();
        try {
            JsonNode root = mapper.readTree(json);
            JsonNode states = root.get("states");
            if (states == null || !states.isArray()) {
                return result;
            }
            for (JsonNode row : states) {
                AircraftState state = AircraftState.fromStateArray(row);
                if (state != null) {
                    result.add(state);
                }
            }
        } catch (Exception e) {
            System.err.println("[OpenSky] failed to parse response: " + e.getMessage());
        }
        return result;
    }
}

File: python/flightradar/opensky_client.py
"""Fetches live aircraft states from the OpenSky Network REST API (``/states/all``).

Used anonymously (no account / OAuth), which is the simplest path but is subject to stricter rate
limits. Any failure (HTTP 429/5xx, timeout, parse error) yields an empty list rather than an
exception, so the UI keeps running and simply skips that update.
"""

import json
from dataclasses import dataclass
from urllib.parse import urlencode
from urllib.request import Request, urlopen

from .aircraft_state import AircraftState

_BASE_URL = "https://opensky-network.org/api/states/all"
_USER_AGENT = "flightradar-python/1.0 (personal demo)"
_TIMEOUT_SECONDS = 10


@dataclass(frozen=True)
class BoundingBox:
    """Bounding box (lamin, lomin, lamax, lomax) in degrees, used to scope and shrink the query."""
    la_min: float
    lo_min: float
    la_max: float
    lo_max: float


#: Roughly continental Europe -- a sensible, traffic-dense default.
EUROPE = BoundingBox(35.0, -12.0, 60.0, 30.0)

#: The whole planet. Heavier payload and higher rate-limit cost; use with care.
WORLD = BoundingBox(-90.0, -180.0, 90.0, 180.0)


class OpenSkyClient:
    def __init__(self, bounding_box=EUROPE):
        self._bbox = bounding_box

    def fetch_states(self):
        """Fetch the current aircraft states within the configured bounding box.

        Returns the live states, or an empty list on any error (never ``None``).
        """
        try:
            request = Request(self._build_url(), headers={"User-Agent": _USER_AGENT})
            with urlopen(request, timeout=_TIMEOUT_SECONDS) as response:
                status = getattr(response, "status", response.getcode())
                if status != 200:
                    print(f"[OpenSky] HTTP {status} -- skipping this update.")
                    return []
                body = response.read().decode("utf-8")
            return self.parse_states(body)
        except Exception as exc:  # noqa: BLE001 - never let a bad fetch kill the app
            print(f"[OpenSky] request failed: {exc} -- skipping this update.")
            return []

    def _build_url(self):
        query = urlencode({
            "lamin": self._bbox.la_min,
            "lomin": self._bbox.lo_min,
            "lamax": self._bbox.la_max,
            "lomax": self._bbox.lo_max,
        })
        return f"{_BASE_URL}?{query}"

    @staticmethod
    def parse_states(body):
        """Parse an OpenSky ``/states/all`` response body into aircraft states.

        Rows without a usable position are skipped. Static so it can be unit tested without HTTP.
        """
        result = []
        try:
            root = json.loads(body)
            states = root.get("states")
            if not isinstance(states, list):
                return result
            for row in states:
                state = AircraftState.from_state_array(row)
                if state is not None:
                    result.append(state)
        except Exception as exc:  # noqa: BLE001
            print(f"[OpenSky] failed to parse response: {exc}")
        return result
